src/hazardous/aead/streaming.rs

```diff
@@ -0,0 +1,1390 @@
+//!
+//!	let mut ctx_enc =
+//! 	StreamXChaCha20Poly1305::new(&secret_key, &nonce);
+//!
+//! // Encrypt and place tag + ciphertext + mac in dst_out_ct
+//!	ctx_enc.seal_chunk(
+//!		message,
+//!		Some(ad),
+//!		&mut dst_out_ct,
+//!		StreamTag::MESSAGE,
+//!	)?;
+//!
+//! let mut ctx_dec =
+//!		StreamXChaCha20Poly1305::new(&secret_key, &nonce);
+//!
+//! // Decrypt and save the tag the message was encrypted with.
+//! let tag = ctx_dec.open_chunk(&dst_out_ct, Some(ad), &mut dst_out_pt)?;
+//!
+//!	assert_eq!(tag, StreamTag::MESSAGE);
+//!	assert_eq!(dst_out_pt.as_ref(), message);
+//! # Ok::<(), orion::errors::UnknownCryptoError>(())
+//! ```
+//! [`SecretKey::generate()`]: ../../stream/chacha20/struct.SecretKey.html
+//! [`Nonce::generate()`]: ../../stream/xchacha20/struct.Nonce.html
+//! [`StreamTag`]: enum.StreamTag.html
+//! [`ABYTES`]: constant.ABYTES.html
+
+use crate::errors::UnknownCryptoError;
+use crate::hazardous::aead::chacha20poly1305::{padding, poly1305_key_gen};
+use crate::hazardous::mac::poly1305::{Poly1305, Tag as Poly1305Tag, POLY1305_OUTSIZE};
+pub use crate::hazardous::stream::chacha20::SecretKey;
+use crate::hazardous::stream::chacha20::{
+	encrypt as chacha20_enc, encrypt_in_place as chacha20_xor_stream, Nonce as IETFNonce,
+	CHACHA_BLOCKSIZE, CHACHA_KEYSIZE, HCHACHA_NONCESIZE, IETF_CHACHA_NONCESIZE,
+};
+use crate::hazardous::stream::xchacha20::subkey_and_nonce;
+pub use crate::hazardous::stream::xchacha20::Nonce;
+use core::convert::TryFrom;
+use subtle::ConstantTimeEq;
+
+#[derive(Debug)]
+/// Tag that indicates the type of message.
+pub enum StreamTag {
+	/// A message with no special meaning.
+	MESSAGE,
+	/// Marks that the message is the end of a set of messages. Allows the decrypting site to
+	/// start working with this data.
+	PUSH,
+	/// Derives a new secret key and forgets the one used for earlier encryption/decryption
+	/// operations.
+	REKEY,
+	/// Indicates the end of a stream. Also does a rekey.
+	FINISH,
+}
+
+impl StreamTag {
+	#[inline(always)]
+	/// Return the tag as a byte.
+	pub fn as_byte(&self) -> u8 {
+		match *self {
+			StreamTag::MESSAGE => 0b0000_0000,
+			StreamTag::PUSH => 0b0000_0001,
+			StreamTag::REKEY => 0b0000_0010,
+			StreamTag::FINISH => 0b0000_0011, // StreamTag::PUSH.as_byte() | StreamTag::REKEY.as_bytes()
+		}
+	}
+}
+
+impl TryFrom<u8> for StreamTag {
+	type Error = UnknownCryptoError;
+
+	fn try_from(byte: u8) -> Result<Self, Self::Error> {
+		match byte {
+			0b0000_0000 => Ok(Self::MESSAGE),
+			0b0000_0001 => Ok(Self::PUSH),
+			0b0000_0010 => Ok(Self::REKEY),
+			0b0000_0011 => Ok(Self::FINISH),
+			_ => Err(UnknownCryptoError),
+		}
+	}
+}
+
+impl core::cmp::PartialEq<StreamTag> for StreamTag {
+	fn eq(&self, other: &StreamTag) -> bool {
+		(self.as_byte().ct_eq(&other.as_byte())).into()
+	}
+}
+
+/// The size of the internal counter.
+const COUNTERBYTES: usize = 4;
+/// The size of the internal nonce.
+const INONCEBYTES: usize = 8;
+/// The size of a Stream.
```

Comment:
The comment here is wrong, its the size of a Tag not Stream

Reply:
Good catch.